### PR TITLE
Add support for testbot infrastructure

### PR DESCRIPTION
**Pull request by telphan**
Signed-off-by: Theodor Gherzan <theodor@resin.io>

### CODE REVIEW

File: lib/workers/testbot.js [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+      portId: HW_SERIAL5,
+      baud: 9600
+    })
+  }
+
+  sendCommand (command, a_ = 0, b_ = 0) {
+    this.board.serialWrite(HW_SERIAL5, [ command, a_, b_ ])
+  }
+
+  async connectSdToTarget (sw) {
```
Comment (jviotti) [inline lib/workers/testbot.js]: I wonder if "connect" is the best verb here. Maybe `switchSDCardToTarget`?

File: lib/workers/testbot.js [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+    return Bluebird.delay(2000)
+  }
+
+  powerTarget (sw) {
+    sw ? this.sendCommand(ENABLE_VOUT_SW) : this.sendCommand(DISABLE_VOUT_SW)
+  }
+
+  async flash (image) {
+    await this.connectSdToTarget(false)
```
Comment (jviotti) [inline lib/workers/testbot.js]: This looks a bit weird. Maybe we should have two functions like:

- `switchSDCardToTarget`
- `switchSDCardToHost`

?

File: lib/workers/testbot.js [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+    if (!drive) {
+      throw new Error(`Cannot find ${destination}`)
+    }
+
+    await mountutils.unmountDiskAsync(drive.device)
+    const driveFileDescriptor = await fs.openAsync(drive.raw, 'rs+')
+    const imageSize = await fs.statAsync(image).get('size')
+
+    const emitter = imageWrite.write({
```
Comment (jviotti) [inline lib/workers/testbot.js]: @zvin Has been making a lot of progress with the Etcher SDK (https://github.com/resin-io-modules/etcher-sdk), so I believe we can start using this module, and abstract away a lot of this code.

File: lib/workers/testbot.js [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+  async ready () {
+    await new Bluebird((resolve, reject) => {
+      this.board.once('ready', resolve)
+      this.board.once('error', reject)
+    })
+
+    this.board.pinMode(LED_PIN, this.board.MODES.OUTPUT)
+
+    // Power managment configuration
```
Comment (jviotti) [inline lib/workers/testbot.js]: Can you explain, using comments, why this works?

File: lib/workers/testbot.js [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    this.board.pinMode(USB_MUX_SEL_PIN, this.board.MODES.OUTPUT)
+    this.board.digitalWrite(USB_MUX_SEL_PIN, this.board.LOW)
+  }
+
+  async on () {
+    console.log('Switching SD card to device...')
+    await this.connectSdToTarget(true)
+
+    console.log('Switching testbot on...')
+    this.powerTarget(true)
```
Comment (jviotti) [inline lib/workers/testbot.js]: The booleans here look really weird. Maybe we should also have `powerTarget` and `shutdownTarget` or something like that?

File: lib/workers/testbot.js [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+const ENABLE_FAULTRST = 0x10
+const SD_RESET_ENABLE = 0x12
+const SD_RESET_DISABLE = 0x13
+
+const LED_PIN = 13
+const HW_SERIAL5 = 5
+const SD_MUX_SEL_PIN = 28
+const USB_MUX_SEL_PIN = 29
+
+const getDevInterface = async (devicePath) => {
```
[the remark this answers is not in the document]
Comment (telphan) [reply lib/workers/testbot.js]: I can move the devIdLink code to the caller, but the same logic will be there, I know the function is a bit weird, but I wanted that this logic to be in one place.

File: lib/workers/testbot.js [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+  const devIdLink = '/dev/disk/by-id/usb-Generic_Ultra_HS-SD_MMC_000008264001-0:0'
+
+  if (process.platform === 'linux') {
+    return fs.realpathAsync(devIdLink).catch((error) => {
```
[the remark this answers is not in the document]
Comment (telphan) [reply lib/workers/testbot.js]: But then if want to output the error, I have to catch again.

File: lib/workers/testbot.js [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+
+module.exports = class TestBotWorker {
+  constructor (title, deviceType, options = {}) {
+    this.title = title
+    this.deviceType = deviceType
+    this.options = options
+
+    this.board = new Board(options.devicePath)
+    this.board.serialConfig({
+      portId: HW_SERIAL5,
```
[the remark this answers is not in the document]
Comment (telphan) [reply lib/workers/testbot.js]: I can add whys, but the answer is very simple. That is the serial configuration. There is no much more reason behind it.
In my opinion the why here is very well expressed by: `portId: 5, baud: 9600`

File: lib/workers/testbot.js (hunk 3 above)
Comment (telphan) [reply lib/workers/testbot.js]: I see, maybe it is worth having a different PR for this, as there are the bits of code which use the same logic and I think the refactoring should be done in the same PR.

**Comment by telphan** [comment]
@jviotti addressed comments.

File: lib/workers/testbot.js [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+* Get dev interface of the SD card
+* @function
+* @param {String} devicePath - This is not used if linux is the host OS
+*
+* @returns {String} path to the dev interface
+*/
+
+const getDevInterface = async (devicePath) => {
+  if (process.platform === 'linux') {
+    const devIdLink = '/dev/disk/by-id/usb-Generic_Ultra_HS-SD_MMC_000008264001-0:0'
```
Comment (jviotti) [inline lib/workers/testbot.js]: Can we move this to a `DEVICE_ID_LINK` constant at the top of the file?

File: lib/workers/testbot.js [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+  * Flash SD card with image
+  * @function
+  * @param {String} - path to the image
+  */
+
+  async flash (image) {
+    await this.switchSdToHost()
+
+    // For linux, udev will provide us with a nice id for the testbot
+    const destination = await getDevInterface(this.options.devicePath)
```
Comment (jviotti) [inline lib/workers/testbot.js]: Can we do something like this instead?

```js
const deviceInterface = process.platform === 'linux' ? DEVICE_ID_LINK : this.options.devicePath

try {
    const destination = await fs.realpathAsync(deviceInterface)
} catch (error) {
    throw new Error(`Could not find testbot. Reason: ${error}`)
}
```

And then we simplify the confusing `getDevInterface` function that doesn't use its arguments on the Linux branch?

Comment (telphan) [reply lib/workers/testbot.js]: Forgot about this, fixed now.

File: lib/workers/testbot.js (hunk 9 above)
Comment (telphan) [reply lib/workers/testbot.js]: Moved up

**Pull request merged**